### ISSUE TITLE
PRD: Pipeline Fixes & Attention-Driven Optimal Formulation

### ISSUE THREAD

**Issue by ChHsiching** [ChHsiching/pharmalink-analyzer#44]
## Problem Statement

PharmaLink Analyzer 在当前状态下存在三个核心问题：

1. **Tauri 桌面客户端无法启动**：无论是开发模式 (`cargo tauri dev`) 还是生产构建 (`cargo tauri build`)，Tauri 客户端均无法运行。用户只能通过浏览器访问 `http://localhost:5173`，但项目最终产物应为 Tauri 桌面应用。

2. **表达式精简与优化功能失效**：在表达式推导页面，点击"精简"或"优化"按钮后，界面先显示灰色等待文字，随即显示红色错误"获取历史失败"，无法完成精简或优化操作。根因涉及前端共享错误状态和可能的后端问题。

3. **核心功能缺失——最优配比**：项目的核心目标是"找出成分浓度与药效之间的数学关系式，然后找到最优解"。当前管线止步于模型评估（Step 5），缺少"基于数学表达式和注意力权重找到最佳成分含量组合"的关键输出。没有这个结果，整个分析流程缺乏实际意义。

此外，测试覆盖不足以发现上述问题。前端零测试，后端缺少端到端集成测试和正确性验证测试。训练结果和注意力分析的准确性也未经验证。

## Solution

分四个阶段修复和扩展应用：

**阶段 1: 核心管线诊断与修复** — 先编写集成测试诊断根因，再修复表达式精简/优化 bug 和前端错误处理。

**阶段 2: 注意力驱动最优配比（核心新功能）** — 新增"最优配比"页面（Step 6），基于注意力权重的分层约束采样（而非纯蒙特卡洛），通过 LHS 在注意力约束的搜索空间中找到 Top 20 组最佳成分含量组合。

**阶段 3: 完善自动化测试** — 添加表达式工作流集成测试、训练正确性验证、注意力权重属性验证、符号回归恢复测试、以 Leaf100HDL 为基准的端到端测试。

**阶段 4: Tauri 桌面客户端修复** — 诊断并修复 Tauri dev 编译/运行问题，配置 Python sidecar 以支持生产打包。

### Bug Fixes — Expression Simplify/Optimize

1. As a researcher, I want to click "精简" and see the expression simplified without errors, so that I can get a cleaner formula
2. As a researcher, I want to click "优化" and navigate the Pareto front without errors, so that I can balance complexity and accuracy
3. As a researcher, I want to see the actual error message when an operation fails (not a misleading "获取历史失败"), so that I can understand what went wrong
4. As a researcher, I want to use "撤销" after a failed simplify without the UI getting stuck, so that I can recover gracefully

### Bug Fixes — Tauri Desktop Client

5. As a user, I want to launch the app via `cargo tauri dev` during development, so that I can test the desktop experience
6. As a user, I want to build a standalone desktop app via `cargo tauri build`, so that I can distribute it without requiring a browser

### Optimal Formulation — Attention Constraints

7. As a researcher, I want to see which components are identified as high-impact by the attention mechanism, so that I understand the search constraints
8. As a researcher, I want to see the Layer 1 dosage range constraints for each component (wider for high-attention, tighter for low-attention), so that I understand how the search space is shaped
9. As a researcher, I want to see the Layer 2 pairwise ratio constraints for synergistic component pairs, so that I understand the interaction constraints
10. As a researcher, I want to see which component pairs are synergistic vs antagonistic before starting the formulation search, so that I can verify the attention analysis is reasonable

### Optimal Formulation — Search and Results

11. As a researcher, I want to start the formulation search with one click, so that I get results quickly
12. As a researcher, I want to configure the search direction (maximize or minimize the therapeutic target), so that I can find the best outcome for my research question
13. As a researcher, I want to configure the number of samples and top-K results, so that I can balance speed and thoroughness
14. As a researcher, I want to see the Top 20 optimal formulations in a table with component content values, so that I can identify the best combination
15. As a researcher, I want to see the predicted therapeutic target value for each formulation, so that I know how effective it should be
16. As a researcher, I want to see a radar chart comparing the optimal formulation vs training data mean, so that I can see how the optimal differs from typical samples
17. As a researcher, I want to export the formulation results to CSV, so that I can use them in other tools
18. As a researcher, I want to see how many valid samples were found after filtering (nan/inf removal), so that I can assess search quality
19. As a researcher, I want the search to automatically increase sampling if valid results are insufficient, so that I always get enough results

### Optimal Formulation — Workflow Integration

20. As a researcher, I want the "最优配比" tab to appear as Step 6 after "模型评估", so that the workflow flows naturally
21. As a researcher, I want the "最优配比" tab to be enabled only when an expression exists, so that I don't attempt formulation without a formula
22. As a researcher, I want the formulation to use the current expression (after simplify/optimize), so that the search reflects my latest formula

### Testing — Correctness Verification

23. As a developer, I want an integration test that runs generate → simplify → optimize → history in sequence, so that I can verify the expression pipeline works end-to-end
24. As a developer, I want a test verifying training loss decreases over epochs, so that I know the transformer actually learns
25. As a developer, I want a test verifying attention weight matrix properties (non-negative, correct shape), so that I know the attention extraction is valid
26. As a developer, I want a test using synthetic data from a known expression to verify symbolic regression recovery (R² > 0.9), so that I know gplearn works correctly
27. As a developer, I want an E2E test using Leaf100HDL data that covers the full pipeline from data loading through formulation, so that I can verify everything works with real data
28. As a developer, I want a test verifying evaluation metrics (R², MSE, MAE) match sklearn reference values, so that I know our metric computation is correct

### Domain Terminology (from CONTEXT.md)

- "最优配比 (Optimal Formulation)" — the new feature, NOT "优化" (which means Pareto front navigation)
- "成分含量 (Component Content)" — the measured value of each component in a sample, NOT "浓度"
- "注意力驱动分层约束采样" — the algorithm name for the formulation search
- All terms respect ADR-0001 (magnitude-based attention classification) and ADR-0002 (gplearn for symbolic regression)

### Algorithm: Attention-Driven Layered Constraint Sampling

The formulation search uses a two-layer constraint approach (NOT dimensionality reduction):

- **Layer 1**: Single-component attention constrains dosage range. High-attention components get wider range [p5, p95]; low-attention components get tighter range [p25, p75]. No features are fixed at mean values.
- **Layer 2**: Pairwise interaction attention constrains ratio ranges for synergistic/antagonistic component pairs. Ratio bounds derived from training data percentiles.

The expression from symbolic regression already only contains high-attention features (low-attention features are eliminated during simplification). Therefore, attention-based dimensionality reduction would be redundant — the algorithm constrains sampling ranges instead.

### Synchronous API

The formulation endpoint is synchronous (POST returns result directly), unlike expression generation which is async. 10K LHS samples with expression evaluation completes in milliseconds via `sympy.lambdify` with numpy backend.

### Numerical Safety

Expressions contain division operators that can produce nan/inf. Evaluation uses `numpy.isfinite()` to filter invalid results. If valid samples < top_k, the algorithm doubles sample count and retries (up to 5 iterations, max ~320K samples).

### Optimizer Dependencies

`FormulationService` derives checkpoint access from `ExpressionState.model_id` → `CheckpointResolver.resolve()`. This avoids redundant dependency injection — the expression state already knows which model it belongs to.

### Frontend Error Handling Fix

The shared `error` ref in `useExpression.ts` causes the simplify/optimize error to be overwritten by "获取历史失败" from `fetchHistory`. Fix: only call `fetchHistory` when the primary operation succeeds (check return value before chaining).

### Test-First Diagnosis

Before fixing the simplify/optimize bug, write `test_expression_workflow.py` integration test to diagnose whether the backend is healthy. If test passes → pure frontend bug. If test fails → backend issue visible in test output.

### Training Strategy

Keep per-dataset training (no pretraining across datasets). The transformer is small enough (d_model=64, 2 layers, 4 heads) for 60-119 samples. Data augmentation + K-Fold provides sufficient regularization.

### Tauri Packaging

Diagnose `cargo tauri dev` failure first. For production: use PyInstaller to create a backend binary, configure as Tauri sidecar via `externalBin`. ADR-0002 confirmed this is feasible (gplearn is pure Python, no Julia dependency).

### New API Contract

```
POST /api/v1/formulation/optimize/{expr_id}?direction=maximize&n_samples=10000&top_k=20
  → FormulationResponse (synchronous)

GET /api/v1/formulation/constraints/{expr_id}
  → AttentionConstraints
```

### New Data Models

```
FormulationRequest:
  n_samples: int = 10000
  top_k: int = 20
  direction: "maximize" | "minimize"

OptimalFormulation:
  component_contents: dict[str, float]   # component name → content value
  predicted_value: float
  rank: int

AttentionConstraints:
  active_features: list[str]
  inactive_features: list[str]
  importance_scores: dict[str, float]
  interaction_pairs: list[tuple[str, str, str]]

SamplingStats:
  total_samples: int
  valid_samples: int
  effective_dimensions: int

FormulationResponse:
  formulations: list[OptimalFormulation]
  constraints: AttentionConstraints
  stats: SamplingStats
```

### Frontend Changes

- New view: `FormulationView` as Step 6 with constraint display, results table, radar chart
- New composable: `useFormulation` for API calls and state
- Router: add `/formulation` route
- Workflow: add `formulation` tab enabled when expression exists
- Tab bar: add "最优配比" as step 6 after "模型评估"

### Testing Philosophy

Test external behavior (given input X, output Y), not implementation details. Each deep module has a clear interface that can be tested in isolation. Use synthetic data for unit tests and Leaf100HDL for integration/E2E tests.

### Modules to Test

| Module | Test Type | What to Verify |
|--------|-----------|---------------|
| `AttentionGuidedFormulator` | Unit | Constraint computation correctness, LHS sampling within bounds, nan/inf filtering, adaptive increment when valid < top_k, ranking order |
| `FormulationService` | Integration | Expression state → attention loading → formulator → response serialization |
| `useExpression fix` | Frontend (Vitest) | Simplify failure does NOT call fetchHistory; error ref shows simplify error, not "获取历史失败" |
| `test_expression_workflow` | Integration (diagnostic) | Generate → simplify → optimize → undo → history chain against real TestClient |
| `test_e2e_leaf100hdl` | E2E | Full pipeline with real data: data load → train → attention → expression → simplify → formulation → verify results reasonable |
| `test_training_correctness` | Correctness | Loss decreases, attention matrix non-negative, K-Fold produces valid models |
| `test_symbolic_correctness` | Correctness | Synthetic data from known expression → gplearn recovery → R² > 0.9 |

### Prior Art

- `test_expression_service.py` (286 lines) — existing service-level test pattern
- `test_e2e_workflow.py` (412 lines) — existing cross-service integration test pattern
- `test_symbolic_regressor.py` (287 lines) — gplearn unit test pattern

### Test Execution

- **Fast tests** (unit + mocked integration): run on every commit, <30s
- **Slow tests** (E2E with real data, `@pytest.mark.slow`): run before merge, ~5min
- **Frontend tests** (Vitest): run on every commit, <15s

## Out of Scope

- **Transformer pretraining/finetuning** — keeping per-dataset training strategy
- **Frontend component library** — continuing hand-rolled UI
- **Export to PDF/report** — can be added later
- **User authentication** — single-user desktop app
- **Mobile support** — desktop only
- **Code splitting for large frontend chunks** — tech debt, not blocking
- **ThreadPoolExecutor vs GIL for gplearn** — known limitation, not user-facing

## Further Notes

- This is a Computer Science undergraduate thesis project (毕业设计)
- Thesis title: "基于注意力机制的中药成分关联性分析研究"
- Domain terminology is documented in `CONTEXT.md` (updated with "最优配比" and "成分含量")
- ADR-0001: attention classification is magnitude-based (not sign-based)
- ADR-0002: gplearn replaced PySR; PyInstaller sidecar is feasible
- Design spec: `docs/superpowers/specs/2026-05-13-pipeline-fixes-optimal-combination-design.md`
- Reference mock data: `/home/chhsich/Git/Mine/tcm-homogenizer/docs/mock/` (expression tree manipulation states from scientific software)
- Recommended test dataset: `Leaf100HDL.csv` (60 samples, 21 features including MDA, target=HDL)
- `docs/` is gitignored — do not commit spec files
